hera_pspec/pspecdata.py

```diff
@@ -680,6 +755,90 @@ def set_Ndlys(self, ndlys=None):
                 raise ValueError("Cannot estimate more delays than there are frequency channels")
             self.spw_Ndlys = ndlys
 
+    def cov_q_hat(self,alpha,beta,key1,key2,time_indices):
+        """
+        Compute the un-normalized covariance matrix for q_hat for a given pair
+        of visibility vectors. Returns the following matrix:
+
+        Cov(\hat{q}_a,\hat{q}_b)
+
+        !!!Only supports covariance between same power-spectrum estimates!!!
+        (covariance between pair of baselines with the same pair of baselines)
+        !!!Assumes that both baselines used in power-spectrum estimate
+        !!!have independent noise relizations!!!
+
+        Parameters
+        ----------
+        alpha,beta: integers for covariance index.
+
+        key1,key2: tuples or lists of tuples
+            Tuples containing the indices of the dataset and baselines for the
+            two input datavectors. If a list of tuples is provided, the baselines
+            in the list will be combined with inverse noise weights.
+
+        time_indices, list of indices for times.
+
+
+        Returns
+        -------
+        cov_q_hat, matrix with covariances between un-normalized band powers
+        """
+        ntimes=len(time_indices)
+
+        qc=np.zeros(ntimes,
+                    dtype=complex)
+
+        R1,R2=0,0
+        n1a,n2a,n1b,n2b=0,0,0,0
+        #compute noise covariance matrices. Assume diagonal!
+                    #compute E^alpha and E^beta
+        if isinstance(key1,list):
+            for _key in key1:
+                R1+=self.R(_key)
+                n1a+=np.real(self.dx(_key))**2.
+                n1b+=np.imag(self.dx(_key))**2.
+        else:
+            R1=self.R(key1)
+            n1a=np.real(self.dx(key1))**2.
+            n1b=np.imag(self.dx(key1))**2.
+
+        if isinstance(key2,list):
+            for _key in key2:
+                R2+=self.R(_key)
+                n2a+=np.real(self.dx(_key))**2.
+                n2b+=np.imag(self.dx(_key))**2.
+        else:
+            R2=self.R(key2)
+            n2a=np.real(self.dx(key2)**2.)
+            n2b=np.imag(self.dx(key2)**2.)
+
+        #rearange Noise arrays into diagonals with Ntimes
+        N1a=np.zeros((self.spw_Nfreqs,self.spw_Nfreqs,ntimes),dtype=complex)
+        N1b=np.zeros_like(N1a)
+        N2a=np.zeros_like(N1a)
+        N2b=np.zeros_like(N1a)
+        for t,tind in enumerate(time_indices):
+            N1a[:,:,t]=np.diag(n1a[:,tind])
+            N1b[:,:,t]=np.diag(n1b[:,tind])
+            N2a[:,:,t]=np.diag(n2a[:,tind])
+            N2b[:,:,t]=np.diag(n2b[:,tind])
+
+        Qalpha=self.get_Q_alt(alpha)
+        Qbeta=self.get_Q_alt(beta)
+        Ealpha=np.einsum('ab,bc,cd',R1.T.conj(),Qalpha,R2)
+        Ebeta=np.einsum('ab,bc,cd',R1.T.conj(),Qbeta,R2)
```

Comment:
Since I also need the E matrices for the decorrelation work, I wrote a separate function for returning them. Just so that we don't have repeated code, we should both use the function. (I know this is my fault for writing function and not telling you about it; sorry!)

To avoid git confusion, let's leave this as-is for now, and then do a very quick PR to address this once all the other issues are addressed.